### PR TITLE
planner: fix panic in expression rewriter caused by `asScalar`

### LINKED ISSUE #20552: panic because of slice bounds out of range
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
Drop table if exists t1;
CREATE TABLE t1(a INT, b INT);
INSERT INTO t1 VALUES(1,1),(2,2),(3,3);
ANALYZE TABLE t1;
SELECT * FROM t1 WHERE ISNULL(t1.a IN (SELECT t3.a FROM t1 t3));
select version();
```

### 2. What did you expect to see? (Required)
```
mysql> SELECT * FROM t1 WHERE ISNULL(t1.a IN (SELECT t3.a FROM t1 t3));
Empty set (0.00 sec)
mysql> select version();
+----------------------------------+
| version()                        |
+----------------------------------+
| 5.7.25-TiDB-v4.0.7-40-gc71000046 |
+----------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> SELECT * FROM t1 WHERE ISNULL(t1.a IN (SELECT t3.a FROM t1 t3));
ERROR 1105 (HY000): runtime error: slice bounds out of range [-1:]
mysql> select version();
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
### 4. What is your TiDB version? (Required)

```
master b71b6323fd4d5ebabff0dfa737aaad3c1cbe797f and release-4.0 c71000046cf1c5ade501cc2b14d3b130863f366d
```

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #20552 

Problem Summary:

SQLs like  `SELECT * FROM t1 WHERE ISNULL(t1.a IN (SELECT t3.a FROM t1 t3));` and `SELECT * FROM t1 WHERE CASE (1 IN (SELECT a FROM t1)) WHEN 0 THEN 0 END;` will cause panic.

### What is changed and how it works?

What's Changed:

Set `asScalar` to true in `(*expressionRewriter).Enter()` when meet `FuncCallExpr` or `CaseExpr`.

How it Works:

`FuncCallExpr` and `CaseExpr` need args be put on `ctxStack`, which means that `asScalar` should be `true`, however it wasn't set to `true`.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- no release note

### CODE REVIEW

File: planner/core/expression_rewriter_test.go [hunk 1]
```diff
@@ -307,3 +307,22 @@ func (s *testExpressionRewriterSuite) TestIssue20007(c *C) {
...
+		store.Close()
+	}()
+
+	tk.MustExec("use test;")
+	tk.MustExec("drop table if exists t1, t2;")
+	tk.MustExec("CREATE TABLE t1(a INT, b INT);")
+	tk.MustExec("INSERT INTO t1 VALUES(1,1),(2,2),(3,3);")
+	tk.MustExec("ANALYZE TABLE t1;")
+	tk.MustExec("SELECT * FROM t1 WHERE ISNULL(t1.a IN (SELECT t3.a FROM t1 t3));")
+	tk.MustExec("SELECT * FROM t1 WHERE CASE (1 IN (SELECT a FROM t1)) WHEN 0 THEN 0 END;")
```
Comment (eurekaka) [inline planner/core/expression_rewriter_test.go:327]: Please check the results of these 2 queries as well.

**Comment by time-and-fate** [comment]
Closed it because #19797 has done the same fix.

**Pull request closed without merging**